client/src/components/CostChart.tsx

```diff
@@ -0,0 +1,496 @@
+import { useEffect, useMemo, useRef, useState } from "react";
+
+import type { ResourceCost, WorkspaceResource } from "../types/api";
+
+interface CostChartProps {
+  costData: ResourceCost[];
+  selectedResources: WorkspaceResource[];
+  loadingResources?: string[];
+}
+
+type ViewType = "table" | "chart";
+
+type TimeSeriesEntry = {
+  date: string;
+  [resourceName: string]: string | number;
+};
+
+function ViewSelector({
+  activeView,
+  onViewChange,
+}: {
+  activeView: ViewType;
+  onViewChange: (view: ViewType) => void;
+}) {
+  return (
+    <div className="mb-4 flex space-x-2">
+      {(["chart", "table"] as ViewType[]).map((view) => (
+        <button
+          key={view}
+          className={`rounded px-4 py-2 ${
+            activeView === view
+              ? "bg-blue-500 text-white"
+              : "bg-gray-200 text-gray-700 hover:bg-gray-300"
+          }`}
+          onClick={() => onViewChange(view)}
+        >
+          {view === "table" ? "Table View" : "Chart View"}
+        </button>
+      ))}
+    </div>
+  );
+}
+
+export function CostChart({ costData, selectedResources, loadingResources = [] }: CostChartProps) {
```

Comment:
this is a very big component, can you try to split it into smaller components ?
E.g. perhaps we could have charts and something else out.
400 LoC right now, hard to reason about.